cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerModificationService.java

```diff
@@ -170,47 +174,114 @@ public void cleanupCluster(Telemetry telemetry) throws CloudbreakException {
+            throws ApiException, CloudbreakException {
+        downloadParcel(Collections.singleton(cdhService), parcelResourceApi);
+        distributeParcel(Collections.singleton(cdhService), parcelResourceApi);
+        callUpgradeCdhCommand(cdhService, clustersResourceApi);
+    }
+
+    private Set<ClouderaManagerProduct> getServices(Set<ClusterComponent> components) {
         return components.stream()
-                .filter(clusterComponent -> clusterComponent.getName().equals(com.sequenceiq.cloudbreak.cloud.model.component.StackType.CDH.name()))
+                .filter(clusterComponent -> ComponentType.CDH_PRODUCT_DETAILS.equals(clusterComponent.getComponentType()))
+                .map(this::getClouderaManagerProduct)
+                .collect(Collectors.toSet());
+    }
+
+    private void upgradeNonCdhServices(Set<ClouderaManagerProduct> services, String cdhServiceName, ParcelResourceApi parcelResourceApi)
+            throws CloudbreakException, ApiException {
+        Set<ClouderaManagerProduct> nonCdhServices = getNonCdhServices(services, cdhServiceName);
+        installServices(nonCdhServices, parcelResourceApi);
+    }
+
+    private void installServices(Set<ClouderaManagerProduct> products, ParcelResourceApi parcelResourceApi) throws ApiException, CloudbreakException {
+        downloadParcel(products, parcelResourceApi);
+        distributeParcel(products, parcelResourceApi);
+        activateParcel(products, parcelResourceApi);
+    }
+
+    private ClouderaManagerProduct getCdhService(Set<ClouderaManagerProduct> products) {
+        return products.stream()
+                .filter(product -> product.getName().equals(com.sequenceiq.cloudbreak.cloud.model.component.StackType.CDH.name()))
                 .findFirst()
                 .orElseThrow(() -> new NotFoundException("Runtime component not found!"));
     }
 
+    private Set<ClouderaManagerProduct> getNonCdhServices(Set<ClouderaManagerProduct> products, String cdhProductName) {
+        return products.stream()
+                .filter(product -> !product.getName().equals(cdhProductName))
+                .collect(Collectors.toSet());
+    }
+
+    private ClouderaManagerProduct getClouderaManagerProduct(ClusterComponent clusterComponent) {
+        try {
+            return clusterComponent.getAttributes().get(ClouderaManagerProduct.class);
+        } catch (IOException e) {
+            LOGGER.info("Could not upgrade Cloudera Runtime services", e);
+            throw new ClouderaManagerOperationFailedException(e.getMessage(), e);
+        }
+    }
+
+    private void checkParcelApiAvailability() throws CloudbreakException {
+        clouderaManagerParcelManagementService.checkParcelApiAvailability(stack, apiClient);
+    }
+
+    private void setParcelRepo(Set<ClouderaManagerProduct> products, ClouderaManagerResourceApi clouderaManagerResourceApi) throws ApiException {
```

Comment:
We should call the method argument to services as well. We're mixing the product and services variables and it's hard to track.